k2/csrc/fsa.cu

```diff
@@ -219,82 +292,96 @@ FsaVec FsaVecFromTensor(Tensor &t, bool *error) {
   }
   K2_CHECK_EQ(sizeof(Arc), sizeof(int32_t) * 4);
   int32_t *tensor_data = t.Data<int32_t>();
-  const Arc *arcs = reinterpret_cast<const Arc *>(tensor_data);
-  ContextPtr c = t.Context();
-  const int32_t num_arcs = t.Dim(0);
-  Array1<int32_t> fsa_ids(c, num_arcs + 1);
-  // `fsa_ids` will be the exclusive sum of `tails`.  We will remove the last
-  // element after we get its value.  Note: `fsa_ids` could be viewed as
-  // `row_ids1`.
-  Array1<int32_t> num_arcs_array(c, 1, num_arcs);
-  const int32_t *num_arcs_data = num_arcs_array.Data();
-  IsTail tails(arcs, num_arcs_data);
-  ExclusiveSum(c, num_arcs + 1, tails, fsa_ids.Data());
-
-  int32_t num_fsas = fsa_ids[num_arcs];
-  fsa_ids = fsa_ids.Range(0, num_arcs);
-  int32_t *fsa_ids_data = fsa_ids.Data();
+
+  Array1<Arc> arc_array(t.Dim(0), t.GetRegion(), t.ByteOffset());
+  return FsaFromArray1(arc_array, error);
+}
+
+
+Fsa FsaVecFromArray1(Array1<Arc> &array, bool *error) {
+  const Arc *arcs_data = reinterpret_cast<const Arc *>(array.Data());
+  ContextPtr c = array.Context();
+  const int32_t num_arcs = array.Dim();
+  Array1<int32_t> row_ids12(c, num_arcs + 1);  // maps arc->fsa_id, like
+                                               // row_ids1[row_ids2]
+  IsLastArcOfFsa fsa_tails(num_arcs, arcs_data);
+  ExclusiveSum(c, num_arcs + 1, fsa_tails, row_ids12.Data());
+  int32_t num_fsas = row_ids12[num_arcs];
+  row_ids12 = row_ids12.Range(0, num_arcs);
+
+  int32_t *fsa_ids_data = row_ids12.Data();
 
   // Get the num-states per FSA, including the final-state which must be
   // numbered last.  If the FSA has arcs entering the final state, that will
-  // tell us what the final-state id is.
-  //   num_state_per_fsa has 2 parts, for 2 different methods of finding the
-  //   right FSA (we'll later shorten it after disambiguating).
-  // The very last element has an error flag that we'll set to 0 if there is
-  // an error.
-  Array1<int32_t> num_states_per_fsa(c, 2 * num_fsas + 1, -1);
+  // tell us what the final-state id is.  (that goes in num_states_per_fsa).
+  // If there are no arcs entering the final-state, we let the final state be
+  // (highest numbered state that has arcs leaving it) + 1, so num_states
+  // (highest numbered state that has arcs leaving it) + 2.
+  //
+  // num_states_per_fsa[num_fsas] is an error flag that gets set to 0 on error.
+  Array1<int32_t> num_states_per_fsa(c, num_fsas + 1, -1);
```

Comment:
Line 344
https://github.com/k2-fsa/k2/blob/bbfa753694996d12a3441474b6ecdf0188927e0b/k2/csrc/fsa.cu#L344

has accesses out of bounds on error.